overlord/servicestate/servicestate.go

```diff
@@ -49,18 +49,55 @@ type Instruction struct {
 
 type ServiceActionConflictError struct{ error }
 
+type ResolvedAppInfo struct {
+	snap.AppInfo
+	ExplicitlyRequested bool
+}
+
+func NewResolvedAppInfo(app *snap.AppInfo) *ResolvedAppInfo {
+	resolvedAppInfo := new(ResolvedAppInfo)
+	resolvedAppInfo.AppInfo = *app
+	return resolvedAppInfo
+}
+
+type ResolvedAppInfos []*ResolvedAppInfo
+
+func NewResolvedAppInfos(appInfos []*snap.AppInfo) ResolvedAppInfos {
+	resolvedAppInfos := make(ResolvedAppInfos, 0, len(appInfos))
+	for _, app := range appInfos {
+		resolvedAppInfos = append(resolvedAppInfos, NewResolvedAppInfo(app))
+	}
+	return resolvedAppInfos
+}
+
+func (a ResolvedAppInfos) Len() int      { return len(a) }
+func (a ResolvedAppInfos) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a ResolvedAppInfos) Less(i, j int) bool {
+	iName := a[i].Snap.InstanceName()
+	jName := a[j].Snap.InstanceName()
+	if iName == jName {
+		return a[i].Name < a[j].Name
+	}
+	return iName < jName
+}
+
 // serviceControlTs creates "service-control" task for every snap derived from appInfos.
-func serviceControlTs(st *state.State, appInfos []*snap.AppInfo, inst *Instruction) (*state.TaskSet, error) {
+func serviceControlTs(st *state.State, appInfos []*ResolvedAppInfo, inst *Instruction) (*state.TaskSet, error) {
 	servicesBySnap := make(map[string][]string, len(appInfos))
 	sortedNames := make([]string, 0, len(appInfos))
+	actionIsOnSnap := make(map[string]bool)
 
 	// group services by snap, we need to create one task for every affected snap
 	for _, app := range appInfos {
 		snapName := app.Snap.InstanceName()
 		if _, ok := servicesBySnap[snapName]; !ok {
 			sortedNames = append(sortedNames, snapName)
+			actionIsOnSnap[snapName] = true
 		}
 		servicesBySnap[snapName] = append(servicesBySnap[snapName], app.Name)
+		if app.ExplicitlyRequested {
+			actionIsOnSnap[snapName] = false
+		}
```

Comment:
We need to also handle an input like this, even if a bit silly:
`snap restart foosnap foosnap.barservice`

(which means snap was given first, then a specific service of the same snap), I'm not sure what will be the outcome atm.